src/extensions/agent_modules/cloud_communication/iothub_client/module/src/iothub_client_module.c

```diff
@@ -0,0 +1,694 @@
+        if (!ADUC_SetDiagnosticsDeviceNameFromConnectionString(info.connectionString))
+        {
+            Log_Error("Setting DiagnosticsDeviceName failed");
+            goto done;
+        }
+
+        if (!IoTHub_CommunicationManager_Init(
+                &g_iotHubClientHandle,
+                ADUC_PnPDeviceTwin_Callback,
+                ADUC_PnP_Components_HandleRefresh,
+                &g_iotHubInitiatedPnPPropertyChangeContext))
+        {
+            Log_Error("IoTHub_CommunicationManager_Init failed");
+            goto done;
+        }
+    }
+
+    // TODO: nox-msft - Instead of passing launchArgs, we should configure agent behaviors (e.g. IoTHub logging)
+    // via a configuration file. This will allow us to remove the launchArgs parameter from this function.
+    if (!ADUC_PnP_Components_Create(g_iotHubClientHandle, 0, NULL))
+    {
+        Log_Error("ADUC_PnP_Components_Create failed");
+        goto done;
+    }
+
+    // The connection string is valid (IoT hub connection successful) and we are ready for further processing.
+    // Send connection string to DO SDK for it to discover the Edge gateway if present.
+    if (ConnectionStringUtils_IsNestedEdge(info.connectionString))
+    {
+        result = ExtensionManager_InitializeContentDownloader(info.connectionString);
+    }
+    else
+    {
+        result = ExtensionManager_InitializeContentDownloader(NULL /*initializeData*/);
+    }
+
+#ifdef ADUC_COMMAND_HELPER_H
+    if (InitializeCommandListenerThread())
+    {
+        RegisterCommand(&redoUpdateCommand);
+    }
+    else
+    {
+        Log_Error(
+            "Cannot initialize the command listener thread. Running another instance of DU Agent with --command will not work correctly.");
+        // Note: even though we can't create command listener here, we need to ensure that
+        // the agent stay alive and connected to the IoT hub.
+    }
+#endif // #ifdef ADUC_COMMAND_HELPER_H
+
+    if (IsAducResultCodeFailure(result.ResultCode))
+    {
+        // Since it is nested edge and if DO fails to accept the connection string, then we go ahead and
+        // fail the startup.
+        Log_Error("Failed to set DO connection string in Nested Edge scenario, result: 0x%08x", result.ResultCode);
+        goto done;
+    }
+
+    result.ResultCode = ADUC_GeneralResult_Success;
+    result.ExtendedResultCode = 0;
+
+done:
+
+    ADUC_ConnectionInfo_DeAlloc(&info);
+    return result.ResultCode == ADUC_GeneralResult_Success ? 0 : -1;
+}
+
+/**
+ * @brief
+ *
+ * @param module
+ *
+ * @return ADUC_Result
+ */
+int IoTHubClientModule_Deinitialize(ADUC_AGENT_MODULE_HANDLE module)
+{
+    Log_Info("Deinitialize");
+    ADUC_D2C_Messaging_Uninit();
+#ifdef ADUC_COMMAND_HELPER_H
+    UninitializeCommandListenerThread();
+#endif
+    ADUC_PnP_Components_Destroy();
+    IoTHub_CommunicationManager_Deinit();
+    DiagnosticsComponent_DestroyDeviceName();
+    ADUC_Logging_Uninit();
+    // TODO: nox-msft Unload all extension
```

Comment:
Could you add work item for this to track if not already done in future PRs or other branch?

Reply:
Will be take care of in the next PR branch.